lua/entities/gmod_wire_expression2/core/array.lua

```diff
@@ -166,63 +164,71 @@ registerCallback( "postinit", function()
-			end)
+			end, nil, nil, NO_LEGACY)
 
 			--------------------------------------------------------------------------------
 			-- Pop functions
 			-- Removes and returns the last value in the array.
 			--------------------------------------------------------------------------------
 			registerFunction( "pop" .. nameupperfirst, "r:", id, function(self,args)
-				local op1 = args[2]
-				local array = op1[1](self,op1)
-				if (!array) then return fixDefault( default ) end
+				local array = args[1]
+				if not array then return fixDefault(default) end
 				return getter( self, array, #array, true )
-			end)
+			end, nil, nil, NO_LEGACY)
 
 			--------------------------------------------------------------------------------
 			-- Unshift functions
 			-- Inserts the value at the beginning of the array. Subsequent values are moved up to compensate.
 			--------------------------------------------------------------------------------
 			registerFunction( "unshift" .. nameupperfirst, "r:" .. id, id, function(self,args)
-				local op1, op2 = args[2], args[3]
-				local array, value = op1[1](self,op1), op2[1](self,op2)
+				local array, value = args[1], args[2]
 				return setter( self, array, 1, value, true )
-			end)
+			end, nil, nil, NO_LEGACY)
 
 			--------------------------------------------------------------------------------
 			-- Shift functions
 			-- Removes and returns the first value of the array. Subsequent values are moved down to compensate.
 			--------------------------------------------------------------------------------
 			registerFunction( "shift" .. nameupperfirst, "r:", id, function(self,args)
-				local op1 = args[2]
-				local array = op1[1](self,op1)
-				if (!array) then return fixDefault( default ) end
+				local array = args[1]
+				if not array then return fixDefault(default) end
 				return getter( self, array, 1, true )
-			end)
+			end, nil, nil, NO_LEGACY)
 
 			--------------------------------------------------------------------------------
 			-- Remove functions
 			-- Removes and returns the specified value of the array. Subsequent values are moved down to compensate.
 			--------------------------------------------------------------------------------
 			registerFunction( "remove" .. nameupperfirst, "r:n", id, function(self,args)
-				local op1, op2 = args[2], args[3]
-				local array, index = op1[1](self,op1), op2[1](self,op2)
-				if (!array or !index) then return fixDefault( default ) end
+				local array, index = args[1], args[2]
+				if not array or not index then return fixDefault(default) end
 				return getter( self, array, index, true )
-			end)
+			end, nil, nil, NO_LEGACY)
+
+			-- indexOf
+
+			registerFunction("indexOf", "r:" .. id, "n", function(self, args)
+				local arr, val = args[1], args[2]
+				for i, j in ipairs(arr) do
+					if j == val then
+						self.prf = self.prf + i
+						return i
+					end
+				end
+				self.prf = self.prf + #arr
```

Comment:
Shouldn't need this with the above prf increment.

I also think this should just be changed to a single `#arr` variable, then add that much prf and loop through that with a manual number loop.

Reply:
> Also need to make this not just add 0 prf for zero length arrays
It still adds ops from the base cost of the function.

Reply:
Also, I don't understand what you mean. Do you mean using `for i = 1, #arr`? I would think that would be less optimal, as `ipairs` is usually better than a normal for loop.

Reply:
Where'd you get that idea from? Afaik `ipairs` only jits on 64 bit. And either way, it's wasted effort for the non-contained case when you loop through the entire array and then have to loop through it again, but in C with the length operator.

Reply:
Huh, I guess I just misremembered, my bad.

Reply:
I'm partially wrong, luajit apparently does cache the length operator? But PUC lua doesn't. It's weird cause my json library got a pretty significant speedup from passing it.

So it's probably not specifically that the length is stored but that it caches any pure operations inside a scope into variables.

Either way it's just better to play it safe.